File: sdk/go/service_registry_test.go
package sdk

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"testing"

	"github.com/hyperledger/fabric-gateway/pkg/client"
	"github.com/nexus-lab/iot-service-blockchain/common"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/mock"
	"github.com/stretchr/testify/suite"
)

type ServiceRegistryTestSuite struct {
	suite.Suite
}

func (s *ServiceRegistryTestSuite) TestRegister() {
	contract := new(MockContract)
	serviceRegistry := &ServiceRegistry{contract}

	service := &common.Service{Name: "service1"}
	data, _ := service.Serialize()
	contract.On("SubmitTransaction", "Register", string(data)).Return(nil, nil)

	err := serviceRegistry.Register(service)
	assert.Nil(s.T(), err, "should return no error")

	err = serviceRegistry.Register(nil)
	assert.Error(s.T(), err, "should return error if input is null")

	service = &common.Service{Name: "service2"}
	data, _ = service.Serialize()
	contract.On("SubmitTransaction", "Register", string(data)).Return(nil, errors.New(""))

	err = serviceRegistry.Register(service)
	assert.Error(s.T(), err, "should return error when sdk or smart contract fails")
}

func (s *ServiceRegistryTestSuite) TestGet() {
	contract := new(MockContract)
	serviceRegistry := &ServiceRegistry{contract}

	expected := new(common.Service)
	data, _ := expected.Serialize()
	contract.On("SubmitTransaction", "Get", "org1", "device1", "service1").Return(data, nil)

	actual, err := serviceRegistry.Get("org1", "device1", "service1")
	assert.Equal(s.T(), expected, actual, "should return correct service")
	assert.Nil(s.T(), err, "should return no error")

	contract.On("SubmitTransaction", "Get", "org2", "device2", "service2").Return(nil, new(common.NotFoundError))

	actual, err = serviceRegistry.Get("org2", "device2", "service2")
	assert.Nil(s.T(), actual, "should return no service")
	assert.IsType(s.T(), new(common.NotFoundError), err, "should return not found error")

	contract.On("SubmitTransaction", "Get", "org3", "device3", "service3").Return(nil, errors.New(""))

	_, err = serviceRegistry.Get("org3", "device3", "service3")
	assert.Error(s.T(), err, "should return error when sdk or smart contract fails")
}

func (s *ServiceRegistryTestSuite) TestGetAll() {
	contract := new(MockContract)
	serviceRegistry := &ServiceRegistry{contract}

	expected := []*common.Service{new(common.Service), new(common.Service)}
	data, _ := json.Marshal(expected)
	contract.On("SubmitTransaction", "GetAll", "org1", "device1").Return(data, nil)

	actual, err := serviceRegistry.GetAll("org1", "device1")
	assert.Equal(s.T(), expected, actual, "should return correct services")
	assert.Nil(s.T(), err, "should return no error")

	contract.On("SubmitTransaction", "GetAll", "org2", "device2").Return([]byte("[]"), nil)

	actual, err = serviceRegistry.GetAll("org2", "device2")
	assert.Zero(s.T(), len(actual), "should return no service")
	assert.Nil(s.T(), err, "should return no error")

	contract.On("SubmitTransaction", "GetAll", "org3", "device3").Return(nil, errors.New(""))

	_, err = serviceRegistry.GetAll("org3", "device3")
	assert.Error(s.T(), err, "should return error when sdk or smart contract fails")
}

func (s *ServiceRegistryTestSuite) TestDeregister() {
	contract := new(MockContract)
	serviceRegistry := &ServiceRegistry{contract}

	service := &common.Service{Name: "service1"}
	data, _ := service.Serialize()
	contract.On("SubmitTransaction", "Deregister", string(data)).Return(nil, nil)

	err := serviceRegistry.Deregister(service)
	assert.Nil(s.T(), err, "should return no error")

	err = serviceRegistry.Deregister(nil)
	assert.Error(s.T(), err, "should return error if input is null")

	service = &common.Service{Name: "service2"}
	data, _ = service.Serialize()
	contract.On("SubmitTransaction", "Deregister", string(data)).Return(nil, errors.New(""))

	err = serviceRegistry.Deregister(service)
	assert.Error(s.T(), err, "should return error when sdk or smart contract fails")
}

func (s *ServiceRegistryTestSuite) TestRegisterEvent() {
	contract := new(MockContract)
	serviceRegistry := &ServiceRegistry{contract}

	eventChannel := make(chan *client.ChaincodeEvent)
	go func() {
		for i := 0; i < 5; i++ {
			data, _ := (&common.Service{Name: fmt.Sprintf("service%d", i)}).Serialize()
			eventChannel <- &client.ChaincodeEvent{
				EventName: fmt.Sprintf("service://org%d/device%d/service%d/register", i, i, i),
				Payload:   data,
			}
		}
	}()

	var cancelFunc context.CancelFunc = func() {
		close(eventChannel)
	}

	contract.On("RegisterEvent", mock.Anything).Return(eventChannel, cancelFunc, nil)

	source, cancel, err := serviceRegistry.RegisterEvent()
	assert.Nil(s.T(), err, "should return no error")
	assert.IsType(s.T(), *new(context.CancelFunc), cancel, "should return correct cancel function")

	for i := 0; i < 5; i++ {
		event := <-source
		assert.Equal(s.T(), "register", event.Action, "should return correct action")
		assert.Equal(s.T(), fmt.Sprintf("org%d", i), event.OrganizationId, "should return correct organization ID")
		assert.Equal(s.T(), fmt.Sprintf("device%d", i), event.DeviceId, "should return correct device ID")
		assert.Equal(s.T(), fmt.Sprintf("service%d", i), event.ServiceName, "should return correct service name")
		assert.IsType(s.T(), new(common.Service), event.Payload, "should return parsed service as event payload")
		assert.Equal(s.T(), fmt.Sprintf("service%d", i), event.Payload.(*common.Service).Name, "should return correct event payload")
	}

	cancel()

	contract = new(MockContract)
	serviceRegistry = &ServiceRegistry{contract}
	contract.On("RegisterEvent", mock.Anything).Return(nil, nil, errors.New(""))

	_, _, err = serviceRegistry.RegisterEvent()
	assert.Error(s.T(), err, "should return error when sdk or smart contract fails")
}

func TestServiceRegistryTestSuite(t *testing.T) {
	suite.Run(t, new(ServiceRegistryTestSuite))
}

File: sdk/javascript/ServiceRegistry.test.ts
import { ChaincodeEvent, CloseableAsyncIterable } from '@hyperledger/fabric-gateway';
import { TextEncoder } from 'util';

import Service from './Service';
import ServiceRegistry from './ServiceRegistry';

const utf8Encoder = new TextEncoder();
const mockContract = () => ({ submitTransaction: jest.fn(), registerEvent: jest.fn() });

test('serviceRegistry.register()', async () => {
  const contract = mockContract();
  const serviceRegistry = new ServiceRegistry(contract);

  const service = new Service('service1', 'device1', 'org1');
  await serviceRegistry.register(service);
  expect(contract.submitTransaction).toHaveBeenCalledWith('Register', service.serialize());

  contract.submitTransaction = jest.fn().mockRejectedValue(new Error());

  await expect(
    serviceRegistry.register(new Service('service2', 'device2', 'org2')),
  ).rejects.toThrow();
});

test('serviceRegistry.get()', async () => {
  const contract = mockContract();
  const serviceRegistry = new ServiceRegistry(contract);

  const service = new Service('service1', 'device1', 'org1');
  contract.submitTransaction = jest.fn().mockResolvedValue(utf8Encoder.encode(service.serialize()));

  const actual = await serviceRegistry.get('org1', 'device1', 'service1');
  expect(actual.serialize()).toEqual(service.serialize());
  expect(contract.submitTransaction).toHaveBeenCalledWith('Get', 'org1', 'device1', 'service1');

  contract.submitTransaction = jest.fn().mockRejectedValue(new Error());

  await expect(serviceRegistry.get('org2', 'device2', 'service2')).rejects.toThrow();
});

test('serviceRegistry.getAll()', async () => {
  const contract = mockContract();
  const serviceRegistry = new ServiceRegistry(contract);

  const expected = [
    new Service('service1', 'device1', 'org1'),
    new Service('service2', 'device2', 'org2'),
  ];
  contract.submitTransaction = jest
    .fn()
    .mockResolvedValue(
      utf8Encoder.encode(JSON.stringify(expected.map((service) => service.toObject()))),
    );

  const actual = await serviceRegistry.getAll('org1', 'device1');
  for (let i = 0; i < 2; i++) {
    expect(actual[i].serialize()).toEqual(expected[i].serialize());
  }
  expect(contract.submitTransaction).toHaveBeenCalledWith('GetAll', 'org1', 'device1');

  contract.submitTransaction = jest.fn().mockResolvedValue(utf8Encoder.encode('[]'));

  await expect(serviceRegistry.getAll('org2', 'device2')).resolves.toEqual([]);

  contract.submitTransaction = jest.fn().mockRejectedValue(new Error());

  await expect(serviceRegistry.getAll('org3', 'device3')).rejects.toThrow();
});

test('serviceRegistry.deregister()', async () => {
  const contract = mockContract();
  const serviceRegistry = new ServiceRegistry(contract);

  const service = new Service('service1', 'device1', 'org1');
  await serviceRegistry.deregister(service);
  expect(contract.submitTransaction).toHaveBeenCalledWith('Deregister', service.serialize());

  contract.submitTransaction = jest.fn().mockRejectedValue(new Error());

  await expect(
    serviceRegistry.deregister(new Service('service2', 'device2', 'org2')),
  ).rejects.toThrow();
});

test('serviceRegistry.registerEvent()', async () => {
  const contract = mockContract();
  const serviceRegistry = new ServiceRegistry(contract);

  contract.registerEvent = jest.fn().mockResolvedValue({
    async *[Symbol.asyncIterator]() {
      for (let i = 0; i < 5; i++) {
        const service = new Service(`service${i}`, `device${i}`, `org${i}`);

        yield {
          eventName: `service://org${i}/device${i}/service${i}/register`,
          payload: utf8Encoder.encode(service.serialize()),
        };
      }
    },
    close: jest.fn(),
  } as CloseableAsyncIterable<ChaincodeEvent>);

  const events = await serviceRegistry.registerEvent();

  try {
    let i = 0;
    for await (const event of events) {
      expect(event.action).toBe('register');
      expect(event.organizationId).toBe(`org${i}`);
      expect(event.deviceId).toBe(`device${i}`);
      expect(event.serviceName).toBe(`service${i}`);
      expect((event.payload as Service).deviceId).toBe(`device${i}`);
      expect((event.payload as Service).organizationId).toBe(`org${i}`);
      expect((event.payload as Service).name).toBe(`service${i}`);

      i++;
    }
    expect(i).toBe(5);
  } finally {
    events.close();
  }

  contract.registerEvent = jest.fn().mockRejectedValue(new Error());

  await expect(serviceRegistry.registerEvent()).rejects.toThrow();
});
